### ISSUE TITLE
Utility Suggestion: Allow toggle for serialized creation and analysis of null matrices

### ISSUE THREAD

**Issue by jonpeake** [taddallas/metacom#16]
I've been using the metacom package for an investigation of the effects of sampling scale and resolution on metacommunity structure of marine fishes. I've been working with fairly large datasets, and I noticed that this results in quite a significant memory leak that I seem to have tracked down to the creation of the list of null matrices. While lapply on the full list of null matrices may run slightly more efficiently from a computing time standpoint, from the standpoint of RAM usage a for loop that creates the null matrix, calculates the necessary metric, and outputs the statistic may be more practical for large/sparse matrices. This could be included as either a logical toggle, or even as a numerical or proportional argument that determines a batch size for null matrix creation and calculation (e.g., say I wanted to produce 1000 nulls but I only wanted them produced and analyzed in batches of 10 to reduce the memory requirement, a hybrid lapply/for loop structure could be implemented that calls NullMaker for a set of 10 matrices 100 separate times in a for loop).

**Comment by taddallas** [comment]
@jonpeake This is a good suggestion. It does feel like most of the time in the functions is in the null matrix creation, especially for large and/or sparse matrices. I'll keep this issue open for now, and try to find some time to carve out an implementation sometime next week.

**Comment by jonpeake** [comment]
@taddallas Thank you, I really appreciate it!

**Comment by jonpeake** [comment]
@taddallas Any progress on this? If need be I could try to take a crack at the implementation.

**Comment by taddallas** [comment]
Looking through it now. Sorry for the delay. The way the code is written, I'm a bit worried that creating nulls in batches might add some time. I've started working on a branch called `sequential` that adds a boolean `sequential` argument. I've implemented it for `Coherence` and in the `Metacommunity` function, but not for `Turnover`. I also just realized that the null matrices get created twice in the `Metacommunity` function, so it may be better to work through coherence and then move to Turnover and such. Boundary clumping doesn't use any null matrices, so we're good there. 

I'm not sure this will solve any problem. Seems like we should address that memory leak more directly in the nullMaker or nullmodel functions. It may be that under such constraints as we place on the nulls, that they are a bit tough to make, resulting in lots of computational time.

**Comment by jonpeake** [comment]
Thanks for taking a look and working through a potential solution. I had a feeling that the sequential null processing might add time, but I'm honestly not sure how else to get around the memory leak. If you have an input matrix that's, say, 25 MB, and you're creating 1000 null matrices that are the same size, that's 25 GB of RAM right off the bat even before you start calculating the test statistic. Batch that into 10 sequential batches of a 100 and you're saving over 22 GB of RAM. Seems like there will always be some sort of trade-off there when it comes to time efficiency and memory efficiency (just like there is when running analyses in parallel, for example). Only other thing I can think of would be to use some implementation of the sparse matrix compression tools from the Matrix package when creating the nulls in nullMaker, but I've never worked with sparse matrices as a dedicated class before so I have no idea how much code change (if any) that would require for the calculation of coherence.

**Comment by taddallas** [comment]
Yea, totally fair. I'm jealous of your 25mb matrices. haha. When I wrote this package, I was relatively new to R, and in a mindset of community ecology, where the typical community matrix is pretty small. Sounds like that's not always the case though! Feel free to play around with the functions and tell me if they're buggy. I have automated tests, but I haven't done much testing of them locally. I'll try to think of other alternatives like your sparse matrix idea (perhaps hold the matrices out of memory or something? idk). Thanks for pointing this out! Hopefully we're at some level of solution.

**Comment by jonpeake** [comment]
Haha indeed, I can imagine that not many were thinking about the application of these sorts of methods to such large datasets spread over space and time. I'll play around with some different potential solutions and let you know if anything jumps out as particularly promising.

**Comment by jonpeake** [comment]
Just wanted to update on a bit of testing I've been running on this. Seems like the batched approach is working very nicely for my needs; I batched both coherence and turnover functions using `for` loops and included a `batch` argument in the function to tweak the number of batches. With my test of a 20-batch run of 1000 sims for both coherence and turnover on a 1500x368 matrix, I was able to reduce the maximum memory usage from 11.5 gigs down to 1.7 gigs. Not a completely linear decrease, but a pretty fantastic decrease all the same. What was even more surprising was that there was actually an accompanying time decrease (on my desktop it decreased from about 5.5 hours to 5 hours). I'm honestly not sure what could've caused the decrease in time (perhaps a reduction in overhead from pulling in the large nulls list each time the function calculated the test statistic), but at the very least it appears that in the case of large matrices any additional runtime overhead from serializing is negligible compared to the realized decrease in memory, and there may even be a slight decrease in runtime. I plan to continue to use my edited versions of the functions for my personal use, but if you would like to implement this batched approach into the main R package I would be happy to share my edited functions.
